Allow checking whether optionals provided explicitly

kstrafe opened the pull request:
Allow checking whether optionals provided explicitly

Since boolean optionals default to "off" we have no way to
check if a boolean optional was actually specified on the command line.

The ability to check if a boolean optional has been given is useful in
cases where we source values from a configuration file where we want any
explicitly given arguments to override the configuration.

An example that mostly works is (for plain optionals with an empty default):

        test -f ~/.standard-values-for-this-script && source "$_"
        MYOPT="${MYOPT:-$_arg_myopt}"
        ... # Code using MYOPT

This works since optionals can default to an empty string. We still are
not entirely sure if the optional was actually passed here, but it tends
to be "good enough" in most cases.

In the case where `myopt` is boolean, the above case would always use
"off" if the argument was not specified, thus overriding the sourced
configuration file.

This patch introduces `_supplied_<argname>`, a value that gets set to 1
if an argument is supplied for that option, otherwise it's set to 0.
This must be enabled on a per-argument basis by declaring:

        ARGBASH_INDICATE_SUPPLIED([long arg name])

Where `long arg name` matches the argument name for any
`ARG_OPTIONAL_*`.

This allows us to do the following:

        test -f ~/.standard-values-for-this-script && source "$_"
        if [ "$_supplied_arg_myopt" = 1 ]; then
                MYOPT="$_arg_myopt" # We override
        fi
        ... # Code using MYOPT

kstrafe commented:
Posting this PR to get some feedback, it's why I haven't written any tests. Let me know if this is a good way to implement this. I think this feature is really useful.

matejak commented:
Very interesting, would you mind adding tests? You basically have to modify e.g. https://github.com/matejak/argbash/blob/master/tests/regressiontests/test-simple.m4, adding an option that would print `_optionals`, so test in https://github.com/matejak/argbash/blob/master/tests/regressiontests/make/tests/tests-base.m4#L125 can be extended accordingly.

kstrafe commented:
> Very interesting, would you mind adding tests? You basically have to modify e.g. https://github.com/matejak/argbash/blob/master/tests/regressiontests/test-simple.m4, adding an option that would print `_optionals`, so test in https://github.com/matejak/argbash/blob/master/tests/regressiontests/make/tests/tests-base.m4#L125 can be extended accordingly.
> 
> Also, could you give a shot to documenting this in https://github.com/matejak/argbash/blob/master/doc/guide.rst#using-parsing-results?

@matejak How do I run the tests? I can't find any documentation on this. Running the test defined in the yml doesn't seem to work for me.

kstrafe commented:
> Very interesting, would you mind adding tests? You basically have to modify e.g. https://github.com/matejak/argbash/blob/master/tests/regressiontests/test-simple.m4, adding an option that would print `_optionals`, so test in https://github.com/matejak/argbash/blob/master/tests/regressiontests/make/tests/tests-base.m4#L125 can be extended accordingly.
> 
> Also, could you give a shot to documenting this in https://github.com/matejak/argbash/blob/master/doc/guide.rst#using-parsing-results?

@matejak I have added two tests, the initial implementation would fail on `x*)` and just print `_optionals+=(arg)`, that is now fixed too.

In addition, shellcheck complained, I've inserted a fix around the shift in parse_commandline. This happened with shellcheck version 0.10.0, and yields the following:

```
In ../tests/regressiontests/test-progs.sh line 72:
                                        shift
                                        ^---^ SC2317 (info): Command appears to be unreachable. Check usage (or ignore if invoked indirectly).
```

In addition to that, I've changed all instances of /bin/bash to /usr/bin/env bash (or env -S bash -e for the cases where it was bash -e).

This is because on my system I do not have a /bin/bash, /usr/bin/env bash is supposed to support more systems. See https://askubuntu.com/questions/1402718/is-the-same-usr-bin-env-bash-than-bin-bash

Let me know if you want to include those two extra commits or not, I can remove them if necessary.

matejak commented:
Great, tests have revealed the problem with non-bash shells that don't have arrays. What about an alternative approach that would use e.g. `_arg_<name>_supplied` variable that could be unset or set to `yes` or `1`?

matejak commented:
FTR, this PR fixes: #185

kstrafe commented:
> Great, tests have revealed the problem with non-bash shells that don't have arrays. What about an alternative approach that would use e.g. `_arg_<name>_supplied` variable that could be unset or set to `yes` or `1`? So one would use `if [[ "$_arg_myopt_supplied" = yes ]]; then` - this would even be quite readable.

@matejak 

This sounds like a good idea but can lead to collisions.
If you define `myopt_supplied` in addition to `myopt`, then setting `myopt` will write `_arg_myopt_supplied`. I think we need a different approach.

I've pushed a commit that uses `_supplied_<arg-name>`, so `quiet` would get `_supplied_arg_quiet=1`.

kstrafe commented:
Fixed the shellcheck error in the test.

matejak commented on src/stuff.m4:
```diff
@@ -731,12 +735,14 @@ m4_define([_MAKE_OPTARG_SIMPLE_CASE_SECTION], [m4_do(
 		[bool],
 		[_JOIN_INDENTED(_INDENT_LEVEL_IN_ARGV_CASE_BODY,
 			[[$5="on"]],
+			[_supplied[$5]=1],
```
```
- [_supplied[$5]=1],
+ [[_supplied$5=1]],
```

For the sake of consistent style, please put the whole assignment (and other occurrences as well)  into double-qoutes - the only reason why to single-quote is when the single-quoted part is a macro, which this is not.

matejak commented on tests/regressiontests/Makefile:
```diff
@@ -373,6 +373,46 @@ test-simple: $(TESTDIR)/test-simple.sh
 	ERROR="require exactly 1" 	$(REVERSE) $<
 	$< pos -o 'uf ta' | grep -q 'OPT_S=uf ta,POS_S=pos,'
 	test -z "$(SHELLCHECK)" || $(SHELLCHECK) "$(TESTDIR)/test-simple.sh"
+test-simple: $(TESTDIR)/test-simple.sh
```
Note to self - I have to refresh my memory how the test definitions work, this feels as unnecessary copy-pasting.

matejak commented on doc/guide.rst:
```diff
@@ -691,6 +691,9 @@ The argument name is transliterated like this:
   If (a boolean argument) ``--quiet`` is passed, value of ``_arg_quiet`` is set to ``on``.
   Conversely, if ``--no-quiet`` is passed, value of ``_arg_quiet`` is set to ``off``.
 
+* Provided optional arguments set ``_supplied_<arg-name>``.
```
I suggest explicitly standardizing to "set to 1", and mentioning it here.

kstrafe replied:
Done.

matejak reviewed (changes requested):
Thanks, I see two issues that need addressing.

Generally, I like your solution very much.
I will help you with tests, so please bear with me for a couple of days.

Finally, please add yourself to AUTHORS file, respecting the alphabetical ordering :-)

matejak commented on tests/regressiontests/test-simple.m4:
```diff
@@ -12,7 +13,11 @@
 # sometimes after a basic validation
-echo "OPT_S=$_arg_prefix,POS_S=$_arg_pos_arg,LA=$_arg_la,"
+if [ "$_arg_print_optionals" = on ]; then
+    echo "OPT_S=$_arg_prefix,POS_S=$_arg_pos_arg,LA=$_arg_la,_supplied_arg_prefix=${_supplied_arg_prefix-x},_supplied_arg_print_optionals=${_supplied_arg_print_optionals-x},_supplied_arg_la=${_supplied_arg_la-x}"
```
Here, for testing purposes, you can decide to drop echoing argument values in this branch of `if`, and focus only on printing the line with supplied arguments.
In the test itself, you can call the script with `--print-optionals` and with other args (or not), and then just check that the expected output matches.

matejak commented on tests/regressiontests/make/tests/tests-base.m4:
```diff
@@ -146,6 +146,14 @@ ADD_TEST_BASH([test-simple], [test_simple_body
 	$< pos -o 'uf ta' | grep -q 'OPT_S=uf ta,POS_S=pos,'
```
```suggestion
	$< pos -o 'uf ta' | grep -q 'OPT_S=uf ta,POS_S=pos,'
	$< pos -o 'uf ta' --print-optionals | grep -q '_supplied_arg_prefix=1,_supplied_arg_print_optionals=1,_supplied_arg_la=x'
```

IOW you can add more lines to this test to check whether the expected supplied arguments match or not to what was actually passed.

matejak commented on tests/regressiontests/make/tests/tests-base.m4:
```diff
@@ -146,6 +146,14 @@ ADD_TEST_BASH([test-simple], [test_simple_body
 	$< pos -o 'uf ta' | grep -q 'OPT_S=uf ta,POS_S=pos,'
 ])
 
+ADD_TEST_BASH([test-simple], [test_simple_body
+	$< pos -o 'uf ta' --print-optionals | grep -q 'OPT_S=uf ta,POS_S=pos,LA=,_supplied_arg_prefix=1,_supplied_arg_print_optionals=1,_supplied_arg_la=x'
+])
+
+ADD_TEST_BASH([test-simple], [test_simple_body
+	$< pos -o 'uf ta' --print-optionals --la x | grep -q 'OPT_S=uf ta,POS_S=pos,LA=x,_supplied_arg_prefix=1,_supplied_arg_print_optionals=1,_supplied_arg_la=1'
+])
+
```
And you can take this out - extending that test above enough.
It tests more things at once, but that's not a problem.

matejak reviewed (changes requested):
Great, I have pointed out a test optimization path for the script and also for the Makefile.

I have been thinking about this feature in general, and I came to a conclusion that the feature should be optional, because the target group is small, and having this on by default would generate larger parsing code for everybody.
There is a precedent for that - the `ARG_POSITIONAL_DOUBLEDASH` or `ARGBASH_SET_DELIM` allow you to set properties of the parsing code in a script-centric (not generator-centric) way.
It's not a big deal and I can help you with that - it is matter of defining a macro, and instead of having the literal assignment to the `_supplied_...` variable, invoke that macro that will either assign stuff, or won't do anything.
Do you have any thoughts on this or on the name of the macro?

kstrafe commented:
> Great, I have pointed out a test optimization path for the script and also for the Makefile.
> 
> I have been thinking about this feature in general, and I came to a conclusion that the feature should be optional, because the target group is small, and having this on by default would generate larger parsing code for everybody. There is a precedent for that - the `ARG_POSITIONAL_DOUBLEDASH` or `ARGBASH_SET_DELIM` allow you to set properties of the parsing code in a script-centric (not generator-centric) way. It's not a big deal and I can help you with that - it is matter of defining a macro, and instead of having the literal assignment to the `_supplied_...` variable, invoke that macro that will either assign stuff, or won't do anything. Do you have any thoughts on this or on the name of the macro?

I'm not sure on this compromise, but I agree that this feature may adds a lot of ultimately unused lines. Even if we add an `ARG_RECORD_SUPPLIED`, we may only be interested whether some optionals were supplied, but not all of them.

One alternative is to tag an optional argument so we can choose which optionals to record. For instance, making another ARG: `ARG_OPTIONAL_SINGLE_SUPPLIED`, but I fear that this will become complicated if we ever want to add one or more more similar features to this one, since we have a combinatorial explosion on the types of arguments; we get `ARG_OPTIONAL_SINGLE_SUPPLIED_X` as well as `ARG_OPTIONAL_SINGLE_X`. I don't think that's desirable.

Alternatively, the "set supplied" flag can be added to the argument list for ARG_*, but since our arguments to `ARG_*` are untagged, and we already have a whole bunch of arguments, readability might be reduced. For example:

`ARG_OPTIONAL_BOOLEAN([argument-name-long], [argument-name-short (optional)], [help message (optional)], [default (optional)], [supplied (optional)])`

Not to mention that we must then provide all preceding optionals to make supplied work.

What about a separate `ARG_SUPPLIED([argument-long-name])`? Here, `argument-long-name` must be an existing `ARG_OPTIONAL_*`. This will tell the code generator to generate a `_supplied_arg_<name>=1` line for that specific argument. This seems to me the most scalable method of going about this.

@matejak what are your thoughts on this? Would this be easy to implement with m4?

kstrafe commented:
> Thank you for your points! I also don't like the idea of extending the API with similar commands differing in only one aspect, or extending the amount of parameters. I like the idea of `ARG_SUPPLIED([argument-long-name], [another-argument-long-name], ...)`, and perhaps if called without any arguments, it can even act as a global "supplied on" switch. It should be doable in `m4`, and I will look into this in course of the next week. Perhaps the name can be different - `ARGBASH_INDICATE_SUPPLIED` or something slightly more verbose. In the meantime, you can work on acceptance tests and documentation, and I will then push the implementation as a commit to your PR. Regarding acceptance tests, you can either create additional scripts and therefore also tests focused solely on testing the "supplied" functionality - you create a file `something.m4`, and introduce a test using `ADD_TEST_BASH([something], [test body])`. The framework will take care of generating the respective `.sh` file and of deleting it at the end.

@matejak 
I've pushed an implementation that makes the user specify `ARG_SUPPLIED`, updated documentation, implemented test.
I've not implemented a global ARG_SUPPLIED flag, you have to specify the arguments for each item you wish to supply.

kstrafe commented:
Thinking of squashing commit `[Default-initialize _supplied_arg_<x>=0 for ARGBASH_INDICATE_SUPPLIED](https://github.com/matejak/argbash/pull/187/commits/8bb10c96c163c494eeed10fc4e55350789c098df)` into the previous one, but leaving it out to see if the solution is acceptable.

matejak commented on src/stuff.m4:
```diff
@@ -1243,7 +1243,10 @@ m4_define([_MAKE_DEFAULTS_OPTIONAL], [m4_do(
 			[incr], [_arg_varname=m4_expand(_default)_ENDL_],
 			[repeated], [_arg_varname=(_default)_ENDL_],
-			[_arg_varname=_sh_quote(_default)_ENDL_])],
+			[m4_do(
+				[_arg_varname=_sh_quote(_default)_ENDL_],
+				[m4_list_contains([HAVE_SUPPLIED], _argname, [[_supplied]_arg_varname=0]_ENDL_, [])],
+			)])],
```
This is inside a `case` statement, which is probably not what you intended. What you want is to perform this check every time, even for `incr` or `repeated` arguments. So you can basically put `[m4_list_contains([HAVE_SUPPLIED], _argname, [[_supplied]_arg_varname=0]_ENDL_, [])],` after the `m4_case` to the same level of indentation as `m4_case`.

kstrafe replied:
I was trying to come up with a use-case for INCR/REPEATED and don't see a solid use-case. Any ideas? I can add it anyway since we do declare all OPTIONALs are supported.

Also, putting it after m4_case will allow positional arguments to also have a `_supplied_arg_...`, I think we should fail in this case because a _supplied_arg for a non-optional doesn't make sense.

matejak commented on tests/regressiontests/test-simple.m4:
```diff
@@ -17,7 +17,8 @@
 # sometimes after a basic validation
 if [ "$_arg_print_optionals" = on ]; then
-    echo "_supplied_arg_prefix=${_supplied_arg_prefix-x},_supplied_arg_print_optionals=${_supplied_arg_print_optionals-x},_supplied_arg_la=${_supplied_arg_la-x},_supplied_arg_not_supplied=${_supplied_arg_not_supplied-x}"
+    set -u
```
This looks like a debug session artifact.

kstrafe replied:
This is to ensure that the arguments that were marked as supplied exist. If they do not exist then the script will fail. I think we should test this to make sure. Alternatively we can test if the `_supplied_arg_x` are set to zero above the if-statement, and exit 1 if they are not.

matejak commented on src/collectors.m4:
```diff
@@ -409,6 +409,7 @@ m4_define([_DISCARD_VALUES_FOR_ALL_ARGUMENTS], [m4_do(
 	[m4_define([_POSITIONALS_INF], 0)],
 	[m4_define([HAVE_POSITIONAL_VARNUM], 0)],
 	[m4_define([HAVE_DOUBLEDASH], 0)],
+	[m4_define([HAVE_SUPPLIED], [])],
```
This statement is ineffective, as `HAVE_SUPPLIED` is used as an identifier of a list, and it is not a macro - the list itself lives as a macro named differently. There is no `m4_list_create` because it is not needed, so please remove this line.

matejak commented on src/collectors.m4:
```diff
@@ -525,6 +526,15 @@ m4_define([_ARG_POSITIONAL_DOUBLEDASH], [m4_do(
+)])
+
+
+m4_define([_ARGBASH_INDICATE_SUPPLIED], [m4_do(
```
You can define the body of the macro right away without the intermediate `_ARGBASH_INDICATE_SUPPLIED`. In this case, there is no benefit of the additional layer.

kstrafe replied:
I agree. Just following convention already used. Will merge.

matejak commented on src/stuff.m4:
```diff
@@ -444,6 +444,7 @@ dnl $5: Where to get the last value (optional)
 m4_define([_VAL_OPT_ADD_SPACE_WITHOUT_GETOPT_OR_SHORT_OPT], [_JOIN_INDENTED(_INDENT_LEVEL_IN_ARGV_CASE_BODY,
 	[test $[]# -lt 2 && die "Missing value for the optional argument '$_key'." 1],
 	[$3([$1], ["@S|@2"], [$4])],
+	[m4_list_contains([HAVE_SUPPLIED], [$1], [_supplied$4=1], [])],
```
You can define a macro `_INDICATION_OF_SUPPLY` in `stuff.m4` for example close to `_APPEND_WRAPPED_ARGUMENT_TO_ARRAY_SPACE`. With body `m4_list_contains([HAVE_SUPPLIED], [$1], [_supplied$2=1], [])` and call it `_INDICATION_OF_SUPPLY([$1], [$4])`

matejak commented on doc/guide.rst:
```diff
@@ -305,6 +305,15 @@ Special arguments
+
+    ARGBASH_INDICATE_SUPPLIED([long arg name], ...)
+
+  This macro will cause a variable to be set if the argument was explicitly provided. This only works for optional arguments.
```
Please also indicate that the macro takes at least one argument, and subsequent values are treated in the same way as subsequent calls to the macro with one argument. Or something like that, if you can phrase it better.

matejak commented on src/collectors.m4:
```diff
@@ -525,6 +526,15 @@ m4_define([_ARG_POSITIONAL_DOUBLEDASH], [m4_do(
+
+
+m4_define([_ARGBASH_INDICATE_SUPPLIED], [m4_do(
+	[m4_list_append([HAVE_SUPPLIED], $@)],
```
Related to the suggestion to change to a set syntax, that would become `m4_set_add_all ([HAVE_SUPPLIED], $@)`.

kstrafe commented:
Added a commit that fails when we use `ARGBASH_INDICATE_SUPPLIED` on non-optional values.
@matejak could you help me out with unittesting this? I'm not sure how to integrate it into the existing framework here. We could just have a new set of tests that runs `argbash <file>` and then greps for a string in the output.

matejak commented on src/collectors.m4:
```diff
@@ -163,6 +163,7 @@ m4_define([__ADD_OPTIONAL_ARGUMENT], [m4_do(
 	[_FILL_IN_VALUES_FOR_AN_OPTIONAL_ARGUMENT([$1], [$3], _arg_varname, [$5], [$2], [$4])],
 	[m4_popdef([_arg_varname])],
 	[m4_define([_DISTINCT_OPTIONAL_ARGS_COUNT], m4_incr(_DISTINCT_OPTIONAL_ARGS_COUNT))],
+	[m4_set_add([OPTIONAL_SET], [$1])],
```
Please select a more descriptive name s.a. `_ALL_OPTIONAL_ARGUMENTS`

matejak commented on src/utilities.m4:
```diff
@@ -340,3 +340,15 @@ m4_define([UNDERLINE], [m4_do(
 	[m4_if(m4_len([$1]), 0, ,
 		[m4_for(idx, 1, m4_len([$1]), 1, [$2])])],
 )])
+
+
+dnl
+dnl $1: Subset
+dnl $2: Superset
+dnl $3: Element binding
+dnl $4: If subset element contained in superset
+dnl $5: If subset element not contained in superset
+m4_define([IS_SUBSET_OF], [m4_do(
+	[m4_set_foreach([$1], [$3],
+		[m4_set_contains([$2], $3, [$4], [$5])])],
+)])
```
I propose to modify this so it uses set operations as noted before. The idea of if-else branching and defining a macro named as the caller wants to use in the error message is very clever, and great for unit testing.

kstrafe replied:
I tried this:
```
m4_define([CHECK_SUPPLIED_ARE_OPTIONAL], [m4_do(
	[m4_set_add_all([SUPPLIED_NOT_OPTIONAL], m4_set_difference([HAVE_SUPPLIED], [_ALL_OPTIONAL_ARGUMENTS]))],
	[m4_if(m4_set_size([SUPPLIED_NOT_OPTIONAL]), 0,
		[],
		[m4_fatal([ARGBASH_INDICATE_SUPPLIED: The following arguments are not optional:] m4_set_dump([SUPPLIED_NOT_OPTIONAL], [, ]))])],
)])
```

But this triggers the fatal even though everything is OK. Any pointers?

matejak replied:
This is a bit tricky. As the documentation for the set operation says, there is always a leading comma output, so you need to remove the comma and space, getting `...OPTIONAL]m4_set_difference(...`
You can also use `m4_set_empty` instead of the size query.
This snippet works, and it is generalized to produce a list of offending values for easier unit-testing. In the calling function, the main question is whether the output argument defines a non-empty macro. As the macro contains commas, it is good to wrap it to m4_quote or m4_expand as in the tests.

I took the liberty of drafting a macro and its test that can be put e.g. to `check-utils.m4`

```
dnl $1 (comma-separated) List of items to check
dnl $2 Set to check against
dnl $3 Output parameter, name of string containing offending arguments
m4_define([CHECK_ITEMS_ARE_IN_SET], [m4_do(
        [m4_set_add_all([_temp_set_supplied], $1)],
        [m4_set_add_all([_temp_set_offenders]m4_set_difference([_temp_set_supplied], [$2]))],
        [m4_define([$3], m4_expand([m4_set_dump([_temp_set_offenders], [, ])]))],,                                
        [m4_set_delete([_temp_set_offenders])],
        [m4_set_delete([_temp_set_supplied])],
)])
```
and
```
m4_set_add_all([CHECK_AGAINST], 1,3,5)                    
CHECK_ITEMS_ARE_IN_SET([1,2,3,4,5], [CHECK_AGAINST], [__output])
assert_equals(m4_expand([__output]), [4, 2])              
CHECK_ITEMS_ARE_IN_SET([1,3,5], [CHECK_AGAINST], [__output])                                                        
assert_equals(m4_expand([__output]), [])                                                                            
CHECK_ITEMS_ARE_IN_SET([8], [CHECK_AGAINST], [__output])  
assert_equals(m4_expand([__output]), [8])
CHECK_ITEMS_ARE_IN_SET([], [CHECK_AGAINST], [__output])   
assert_equals(m4_expand([__output]), [])
m4_set_delete([CHECK_AGAINST])
```

kstrafe replied:
I missed that first suggestion where you omitted the comma. What a surprising language. I've pushed a commit that separates fatal errors from the check itself.

matejak commented:
> Added a commit that fails when we use ARGBASH_INDICATE_SUPPLIED on non-optional values.

I have added some pointers toward unittests, and you can also add tests that are supposed to fail generation like here: A [gentest](https://github.com/matejak/argbash/blob/master/tests/regressiontests/gen-test-infinity.m4) testing e.g. [infinitely many arguments](https://github.com/matejak/argbash/blob/master/tests/regressiontests/make/tests/tests-base.m4#L319) that are not supported in Dash uses [gen-test-infinity.m4](https://github.com/matejak/argbash/blob/master/tests/regressiontests/gen-test-infinity.m4) under the hood as the file that shouldn't compile by argbash with posix output.

One more thing before the merge - could you rebase the PR against the current master branch? There are some commits showing up that are already merged.

kstrafe commented:
I've changed the logic for the subset checking and decoupled the fatal so the unittest can test the subset logic.
Furthermore, a gen-test has been added to ensure we fail on all if the argument was not provided.

Additionally, we now also do not allow optional incr, repeated, and action to be tagged as ARGBASH_INDICATE_SUPPLIED; it makes no sense to do so, since we already have the information needed to figure out if an argument was supplied.

matejak reviewed (approved):
Well done, thank you for your heroic effort!
